topdown/eval.go

```diff
@@ -471,18 +471,26 @@ func (e *eval) evalNotPartial(iter evalIterator) error {
 	// dependencies of the negation.
 	unknowns := e.saveSet.Vars(e.caller.bindings)
 
-	// Run partial evaluation, plugging the result and applying copy propagation to
-	// each result. Since the result may require support, push a new query onto the
-	// save stack to avoid mutating the current save query.
-	p := copypropagation.New(unknowns).WithEnsureNonEmptyBody(true).WithCompiler(e.compiler)
+	// Run partial evaluation. Since the result may require support, push a new
+	// query onto the save stack to avoid mutating the current save query. If
+	// shallow inlining is not enabled, run copy propagation to further simplify
+	// the result.
+	var cp *copypropagation.CopyPropagator
+
+	if !e.inliningControl.shallow {
```

Comment:
💭   Seeing these kind of checks around kind of makes me wonder if we would be better served with more specific controls for like "do copy propagation" which gets set depending on some combination of other factors. As-is its just the one factor so no biggie, but eventually with more optimization levels and stuff I could see us wanting to refactor this a little bit.

To clarify, not asking for a change now, just something for us to keep an eye on in the future.